fineract-progressive-loan/src/main/java/org/apache/fineract/portfolio/loanaccount/service/ProgressiveLoanInterestRefundServiceImpl.java

```diff
@@ -0,0 +1,130 @@
+        List<LoanTransaction> withoutSpecificDisbursements = loan.getLoanTransactions().stream()
+                .filter(lt -> !lt.isAccrual() && !lt.isAccrualActivity() && !lt.isReversed()).filter(lt -> {
+                    if (lt.getTypeOf().isDisbursement() && refundFinal.get().compareTo(BigDecimal.ZERO) > 0) {
+                        //
+                        if (lt.getAmount().compareTo(refundFinal.get()) <= 0) {
+                            refundFinal.set(refundFinal.get().subtract(lt.getAmount()));
+                            return false;
+                        }
+                    }
+                    return true;
+                }).toList();
+        List<LoanTransaction> transactionsToReprocess = withoutSpecificDisbursements.stream().map(lt -> {
+            if (lt.getTypeOf().isDisbursement() && refundFinal.get().compareTo(BigDecimal.ZERO) > 0) {
+                refundFinal.set(BigDecimal.ZERO);
+                return new LoanTransaction(lt.getLoan(), lt.getLoan().getOffice(), lt.getTypeOf().getValue(), lt.getDateOf(),
+                        lt.getAmount().subtract(refundFinal.get()), lt.getPrincipalPortion(), lt.getInterestPortion(),
+                        lt.getFeeChargesPortion(), lt.getPenaltyChargesPortion(),
+                        lt.getOverPaymentPortion(lt.getLoan().getCurrency()).getAmount(), lt.isReversed(), lt.getPaymentDetail(),
+                        lt.getExternalId());
+            }
+            return new LoanTransaction(lt.getLoan(), lt.getLoan().getOffice(), lt.getTypeOf().getValue(), lt.getDateOf(), lt.getAmount(),
+                    lt.getPrincipalPortion(), lt.getInterestPortion(), lt.getFeeChargesPortion(), lt.getPenaltyChargesPortion(),
+                    lt.getOverPaymentPortion(lt.getLoan().getCurrency()).getAmount(), lt.isReversed(), lt.getPaymentDetail(),
+                    lt.getExternalId());
+        }).toList();
+
+        List<LoanRepaymentScheduleInstallment> installmentsToReprocess = new java.util.ArrayList<>(
+                loan.getRepaymentScheduleInstallments().stream().filter(i -> !i.isReAged() && !i.isAdditional())
+                        .map(ProgressiveLoanInterestRefundServiceImpl::duplicate).toList());
+
+        return processor.reprocessProgressiveLoanTransactions(loan.getDisbursementDate(), transactionsToReprocess, loan.getCurrency(),
+                installmentsToReprocess, loan.getActiveCharges()).getRight();
+    }
+
+    @Override
+    public boolean canHandle(Loan loan) {
+        return loan != null && loan.isInterestBearing()
+                && (AdvancedPaymentScheduleTransactionProcessor.ADVANCED_PAYMENT_ALLOCATION_STRATEGY
+                        .equalsIgnoreCase(loan.getTransactionProcessingStrategyCode())
+                        || AdvancedPaymentScheduleTransactionProcessor.ADVANCED_PAYMENT_ALLOCATION_STRATEGY_NAME
+                                .equalsIgnoreCase(loan.getTransactionProcessingStrategyName()));
```

Comment:
Is this even possible that the allocation strategy code doesn't match the name?

Reply:
Technically yes, because `AdvancedPaymentTransactionProcessor` has the same logic and I did not found strictier validation.